iroha/src/kura.rs

```diff
@@ -180,126 +189,252 @@ impl Default for Mode {
         if fs::read_dir(path).is_err() {
             fs::create_dir_all(path).wrap_err("Failed to create Block Store directory.")?;
         }
         Ok(BlockStore {
             path: path.to_path_buf(),
+            blocks_per_file,
         })
     }
 
-    fn get_block_filename(block_height: u64) -> String {
-        format!("{}", block_height)
+    /// Returns expected filename of datafile containing block, provided its height
+    async fn get_block_filename(
+        dir: &Path,
+        block_height: NonZeroU64,
+        blocks_per_file: NonZeroU64,
+    ) -> Result<String> {
+        const INITIAL_DATA_FILE: &str = "1";
+        let storage_indices = sorted_storage_files_base_indices(dir)
+            .await
+            .wrap_err("Failed to list storage files")?;
+        match storage_indices.last() {
+            Some(max_base) => {
+                if block_height <= *max_base {
+                    return Err(eyre::eyre!("An attempt to random-access persistent storage (file except for the last one queued)."));
+                }
+                if (block_height.get() - max_base.get()) < blocks_per_file.get() {
+                    return Ok(max_base.to_string());
+                }
+                Ok(block_height.to_string())
+            }
+            None => Ok(String::from(INITIAL_DATA_FILE)),
+        }
     }
 
     /// Get filysystem path for the block at height.
-    pub fn get_block_path(&self, block_height: u64) -> PathBuf {
-        self.path.join(BlockStore::get_block_filename(block_height))
+    ///
+    /// # Errors
+    ///
+    pub async fn get_block_path(&self, block_height: NonZeroU64) -> Result<PathBuf> {
+        let filename =
+            BlockStore::get_block_filename(&self.path, block_height, self.blocks_per_file)
+                .await
+                .wrap_err("Failed to fetch storage name for block")?;
+        Ok(self.path.join(filename))
     }
 
     /// Write block as file onto the disk.
     #[iroha_futures::telemetry_future]
     pub async fn write(&self, block: &VersionedCommittedBlock) -> Result<Hash> {
-        //filename is its height
-        let path = self.get_block_path(block.header().height);
-        let mut file = File::create(path)
+        let path = self
+            .get_block_path(
+                NonZeroU64::new(block.header().height)
+                    .ok_or_else(|| eyre::eyre!("Block height is non-positive."))?,
+            )
+            .await?;
+        let mut file = OpenOptions::new()
+            .create(true)
+            .append(true)
+            .open(path)
             .await
             .wrap_err("Failed to open storage file.")?;
         let hash = block.hash();
-        let serialized_block: Vec<u8> = block.encode_versioned()?;
+        let mut encoded: Vec<u8> = block.encode_versioned()?;
+        let serialized_block = prepend_size(&mut encoded);
         file.write_all(&serialized_block)
```

Comment:
You can write size directly to file and then write encoded block. That should do the job.

Reply:
You dislike another function here?

Reply:
You dislike a new function here?

Reply:
Removed it.